core/src/subgraph/instance.rs

```diff
@@ -90,7 +94,12 @@ where
         data_source: C::DataSource,
         templates: Arc<Vec<C::DataSourceTemplate>>,
         host_metrics: Arc<HostMetrics>,
-    ) -> Result<T::Host, Error> {
+    ) -> Result<Option<T::Host>, Error> {
+        let host_builder = match self.host_builder {
+            None => return Ok(None),
+            Some(ref host_builder) => host_builder,
+        };
+
         let mapping_request_sender = {
             let module_bytes = data_source.runtime();
```

Comment:
What about making this controlled by the data source, so we change `DataSource::runtime` to return an option, and if it returns none then no host is created.